src/Microsoft.ServiceFabric.Actors/Runtime/MigrationOrchestrator.cs

```diff
@@ -0,0 +1,452 @@
+
+                                if (lastAppliedSN != -1)
+                                {
+                                    tasks.Add(worker.StartMigrationWorker(MigrationPhase.Copy, lastAppliedSN, endSN, i, cancellationToken));
+                                }
+                                else
+                                {
+                                    tasks.Add(worker.StartMigrationWorker(MigrationPhase.Copy, startSN, endSN, i, cancellationToken));
+                                }
+                            }
+                        });
+                        await Task.WhenAll(tasks);
+
+                        var lastUpdatedRecord = await this.StartCatchupPhaseWorkload(endSN, cancellationToken);
+                        await this.StartDowntimePhaseWorkload(lastUpdatedRecord, cancellationToken);
+                    }
+                    else
+                    {
+                        var lastUpdatedRecord = await this.StartCatchupPhaseWorkload(endSN, cancellationToken);
+                        await this.StartDowntimePhaseWorkload(lastUpdatedRecord, cancellationToken);
+                    }
+                }
+                else if (phase == MigrationPhase.Catchup)
+                {
+                    ConditionalValue<byte[]> iterationValue, startSNValue;
+                    using (var tx = this.stateProvider.GetStateManager().CreateTransaction())
+                    {
+                        iterationValue = await this.metadataDict.TryGetValueAsync(tx, MigrationConstants.CatchupIterationKey);
+                        startSNValue = await this.metadataDict.TryGetValueAsync(tx, MigrationConstants.CatchupStartSNKey);
+                    }
+
+                    if (!iterationValue.HasValue)
+                    {
+                        long.TryParse(Encoding.ASCII.GetString(startSNValue.Value), out long startSN);
+                        var lastAppliedSN = await this.StartCatchupPhaseWorkload(startSN - 1, cancellationToken);
+                        cancellationToken.ThrowIfCancellationRequested();
+                        await this.StartDowntimePhaseWorkload(lastAppliedSN, cancellationToken);
+                    }
+                    else
+                    {
+                        int.TryParse(Encoding.ASCII.GetString(iterationValue.Value), out int iterationCount);
+                        ConditionalValue<byte[]> lastSNValue, lastAppliedSNValue;
+                        using (var tx = this.stateProvider.GetStateManager().CreateTransaction())
+                        {
+                            lastSNValue = await this.metadataDict.TryGetValueAsync(tx, MigrationConstants.GetCatchupWorkerEndSNKey(iterationCount));
+                            lastAppliedSNValue = await this.metadataDict.TryGetValueAsync(tx, MigrationConstants.GetCatchupWorkerLastAppliedSNKey(iterationCount));
+                        }
+
+                        if (lastAppliedSNValue.HasValue)
```

Comment:
For code readability it would be better if you move the phase specific code to separate private method.

Reply:
Updated